### ISSUE TITLE
Searching for similar tracks using audio embeddings

### ISSUE THREAD

**Issue by edwardgowsmith** [mixxxdj/mixxx#13975]
### Feature Description

I've integrated a search function into my local version of Mixxx which allows for searching for similar tracks using AI audio models.

I was inspired by the website https://cosine.club/, which uses Essentia models (https://essentia.upf.edu/models.html), returning the most similar tracks to an input track based on cosine similarities between the embedding vectors.

Here are some screenshots of the current functionality:

![image](https://github.com/user-attachments/assets/9b044b82-0eec-4622-99ba-665e9909eb78)
![image](https://github.com/user-attachments/assets/8d34c8f2-90bc-4797-9b28-b4df395be22c)

Is this something that would be of interest to people as a feature?
If so, I can work on it further as its own branch.
There's a few big issues as it stands:
- The search returns the 10 most similar tracks in your collection, ordered by artist, not similarity.
- The embeddings are calculated with an external Python script that updates the SQLite database. Should be fairly easy to get this to run within Mixxx, though?

Also, the Essentia models are licensed under CC BY-NC-SA 4.0. Is this a problem?

**Comment by edwardgowsmith** [comment]
I was naively thinking of calling a Python script within Mixxx that would calculate the embeddings -- is that feasible?
I'm not sure about the possibility of running the model in C++. From a little investigation it seems it might be pretty tricky!
Regardless, it might be an interesting project for me to give it a try.

**Comment by Swiftb0y** [comment]
> I was naively thinking of calling a Python script within Mixxx that would calculate the embeddings -- is that feasible?

Yes, but not robust. We don't want to rely on python when running nor do we want to distribute it. A proper solution would need to use (ideally hardware accelerated) C++. The essentia site mentions that some models are available ONNX format. Using ONNX would be the preferred solution. Which model are you using exactly? Is that available in ONNX format? Then you could use https://onnxruntime.ai/ for inferencing. That would be the most performant, most robust and least maintenance-requiring solution.

**Comment by fwcd** [comment]
Super exciting to see this and would love to see a PR! I'm sure we can figure out the licensing issues, in the worst case we'd have to download the model after installation or something

**Comment by Swiftb0y** [comment]
@edwardgowsmith. Fyi if you're looking for a point where to start implementing this in mixxx, take a look at the classes in the `src/analyzer/` directory. This is where you can retrieve a tracks sample data and the database to tie both together.
